pkg/pubsub/adapter/converters/auditlog.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package converters
+
+import (
+	"bytes"
+	"context"
+	errors "errors"
+	"fmt"
+	"log"
+	"reflect"
+	"regexp"
+	"strings"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	cepubsub "github.com/cloudevents/sdk-go/pkg/cloudevents/transport/pubsub"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/google/knative-gcp/pkg/apis/events/v1alpha1"
+	auditpb "google.golang.org/genproto/googleapis/cloud/audit"
+	logpb "google.golang.org/genproto/googleapis/logging/v2"
+)
+
+const (
+	AuditLogAdapterType = "google.auditlog"
+
+	logEntrySchema = "type.googleapis.com/google.logging.v2.LogEntry"
+	loggingSource  = "logging.googleapis.com"
+
+	parentResourcePattern = `^(:?projects|organizations|billingAccounts|folders)/[^/]+`
+
+	serviceNameExtension  = "servicename"
+	methodNameExtension   = "methodname"
+	resourceNameExtension = "resourcename"
+)
+
+var (
+	jsonpbUnmarshaller = jsonpb.Unmarshaler{
+		AllowUnknownFields: true,
+		AnyResolver:        resolver(resolveAnyUnknowns),
+	}
+	jsonpbMarshaler      = jsonpb.Marshaler{}
+	parentResourceRegexp *regexp.Regexp
+)
+
+func init() {
+	var err error
+	if parentResourceRegexp, err = regexp.Compile(parentResourcePattern); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// Resolver function type that can be used to resolve Any fields in a jsonpb.Unmarshaler.
+type resolver func(turl string) (proto.Message, error)
+
+func (r resolver) Resolve(turl string) (proto.Message, error) {
+	return r(turl)
+}
+
+type UnknownMsg empty.Empty
+
+func (m *UnknownMsg) ProtoMessage() {
+	(*empty.Empty)(m).ProtoMessage()
+}
+
+func (m *UnknownMsg) Reset() {
+	(*empty.Empty)(m).Reset()
+}
+
+func (m *UnknownMsg) String() string {
+	return "Unknown message"
+}
+
+// Resolver type which resolves unknown message types to empty.Empty.
+func resolveAnyUnknowns(typeURL string) (proto.Message, error) {
+	// Only the part of typeUrl after the last slash is relevant.
+	mname := typeURL
```

Comment:
An example of the typeURL in the comment would be helpful.

Reply:
I don't think id is meant to be human readable, so I'm not sure this is needed.

Reply:
Done.